sql/grant_tables/privilege_set.go

```diff
@@ -0,0 +1,614 @@
+func (ps PrivilegeSetDatabase) ToSlice() []PrivilegeType {
+	privs := make([]PrivilegeType, len(ps.privs))
+	i := 0
+	for priv := range ps.privs {
+		privs[i] = priv
+		i++
+	}
+	return privs
+}
+
+// ToSortedSlice returns all of the database privileges contained as a slice, sorted by their internal ID.
+func (ps PrivilegeSetDatabase) ToSortedSlice() []PrivilegeType {
+	privs := ps.ToSlice()
+	sort.Slice(privs, func(i, j int) bool {
+		return privs[i] < privs[j]
+	})
+	return privs
+}
+
+// getUseableTbl is used internally to either retrieve an existing table, or create a new one that is returned.
+func (ps PrivilegeSetDatabase) getUseableTbl(tblName string) PrivilegeSetTable {
+	lowerTblName := strings.ToLower(tblName)
+	tblSet, ok := ps.tables[lowerTblName]
+	if !ok {
+		tblSet = PrivilegeSetTable{
+			name:    tblName,
+			privs:   make(map[PrivilegeType]struct{}),
+			columns: make(map[string]PrivilegeSetColumn),
+		}
+		ps.tables[lowerTblName] = tblSet
+	}
+	return tblSet
+}
+
+// merge merges the given set of privileges to the calling set of privileges.
+func (ps PrivilegeSetDatabase) merge(otherPs PrivilegeSetDatabase) {
+	for priv := range otherPs.privs {
+		ps.privs[priv] = struct{}{}
+	}
+	for _, otherTblSet := range otherPs.tables {
+		ps.getUseableTbl(otherTblSet.name).merge(otherTblSet)
+	}
+}
+
+// clear removes all database privileges.
+func (ps PrivilegeSetDatabase) clear() {
+	for priv := range ps.privs {
+		delete(ps.privs, priv)
+	}
+}
+
+// PrivilegeSetTable is a set containing table-level privileges.
+type PrivilegeSetTable struct {
+	name    string
+	privs   map[PrivilegeType]struct{}
+	columns map[string]PrivilegeSetColumn
+}
+
+// Name returns the name of the table that this privilege set belongs to.
+func (ps PrivilegeSetTable) Name() string {
+	return ps.name
+}
+
+// Has returns whether the given table privilege(s) exists.
+func (ps PrivilegeSetTable) Has(privileges ...PrivilegeType) bool {
+	for _, priv := range privileges {
+		if _, ok := ps.privs[priv]; !ok {
+			return false
+		}
+	}
+	return true
+}
+
+// Count returns the number of table privileges.
+func (ps PrivilegeSetTable) Count() int {
+	return len(ps.privs)
+}
+
+// Column returns the set of privileges for the given column. Returns an empty set if the column does not exist.
+func (ps PrivilegeSetTable) Column(colName string) PrivilegeSetColumn {
+	colSet, ok := ps.columns[strings.ToLower(colName)]
+	if ok {
+		return colSet
+	}
+	return PrivilegeSetColumn{name: colName}
+}
+
+// GetColumns returns all columns.
+func (ps PrivilegeSetTable) GetColumns() []PrivilegeSetColumn {
+	colSets := make([]PrivilegeSetColumn, len(ps.columns))
+	i := 0
+	for _, colSet := range ps.columns {
+		colSets[i] = colSet
+		i++
+	}
+	return colSets
+}
+
+// Equals returns whether the given set of privileges is equivalent to the calling set.
+func (ps PrivilegeSetTable) Equals(otherPs PrivilegeSetTable) bool {
```

Comment:
The presence of this makes me suspicious. Why is this needed?

Reply:
This will be passed on to integrators, who may need convenience functions like that.

Reply:
Sometimes I do forget to remove a function that was needed in some pre-PR development, but at least for many of these "clearer" functions, they're often for integrators or they're mirroring other used functions (`Left()` is used and `Right()` isn't, but it makes more sense to have both).